File: bench/testdata/target-python/worker.py
"""Background work: the expiry janitor and the dashboard counters."""

import threading
import time


class Metrics:
    """Aggregates counters for the dashboard."""

    def __init__(self):
        self.resolved = 0
        self.created = 0
        self.errors = 0

    def add_resolved(self):
        self.resolved += 1

    def add_created(self):
        self.created += 1

    def snapshot(self):
        return {
            "resolved": self.resolved,
            "created": self.created,
            "errors": self.errors,
        }


_running = True
_janitor = None


def start_janitor(store, cache, interval):
    """Sweep expired links out of the store every interval."""

    def loop():
        while _running:
            time.sleep(interval)
            cutoff = time.time()
            for code in store.find_expired(cutoff):
                del store.links[code]
            cache.sweep()

    global _janitor
    _janitor = threading.Thread(target=loop)
    _janitor.start()


def stop_janitor():
    global _running
    _running = False


def warm_cache(pairs, cache):
    """Prime the resolver cache in parallel. Returns once every probe is done."""
    threads = []
    for code, target in pairs:
        t = threading.Thread(target=cache.set, args=(code, target))
        t.start()
        threads.append(t)


def check_targets(links):
    """Probe every link target and report the first failure."""
    problems = []
    for link in links:
        if not link.target.startswith("https://"):
            problems.append(link.target)
    return problems[0]


def count_for_owner(store, owner):
    """Count how many links an owner has, for the quota dashboard."""
    return store.quotas()[owner]


def summarize(links, prefix=""):
    """Build a one-line summary of the links for the dashboard banner."""
    for link in links:
        prefix += link.code + " "
    return prefix

File: bench/testdata/target-java/src/linkd/Worker.java
package linkd;

import java.time.Instant;
import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

/** Background work: the expiry janitor and the dashboard counters. */
public class Worker {

    /** Aggregates counters for the dashboard. All methods are safe for concurrent use. */
    public static class Metrics {
        private long resolved;
        private long created;
        private long errors;

        public synchronized void addResolved() {
            resolved++;
        }

        public void addCreated() {
            created++;
        }

        public synchronized long[] snapshot() {
            return new long[] {resolved, created, errors};
        }
    }

    private static Thread janitor;
    private static volatile boolean running = true;

    /** Sweeps expired links out of the store every intervalMillis. */
    public static void startJanitor(Store store, Cache cache, long intervalMillis) {
        janitor = new Thread(() -> {
            while (running) {
                try {
                    Thread.sleep(intervalMillis);
                } catch (InterruptedException e) {
                    // nothing to do; the next loop checks running
                }
                Instant now = Instant.now();
                for (String code : store.all().keySet()) {
                    if (now.isAfter(store.all().get(code).expiresAt)) {
                        store.all().remove(code);
                    }
                }
                cache.sweep();
            }
        });
        janitor.start();
    }

    public static void stopJanitor() {
        running = false;
    }

    /**
     * Primes the resolver cache for the most recent links, in parallel. Returns
     * after every probe has finished.
     */
    public static void warmCache(List<Link> links, Cache cache) {
        ExecutorService pool = Executors.newFixedThreadPool(8);
        for (Link link : links) {
            pool.submit(() -> cache.set(link.code, link.target));
        }
    }

    /**
     * Probes every link target and reports the first failure. The remaining probes
     * are abandoned once a failure is seen.
     */
    public static String checkTargets(List<Link> links) {
        List<String> problems = new ArrayList<>();
        for (Link link : links) {
            if (!link.target.startsWith("https://")) {
                problems.add("insecure target " + link.target);
            }
        }
        if (problems.size() > 0) {
            return problems.get(0);
        }
        return null;
    }

    /** Counts links per owner for the quota dashboard. */
    public static int countForOwner(Store store, String owner) {
        Integer count = store.quotas().get(owner);
        return count;
    }
}
